server/src/test/java/io/druid/client/HttpServerInventoryViewTest.java

```diff
@@ -0,0 +1,334 @@
+        MoreExecutors.sameThreadExecutor(),
+        new ServerView.ServerRemovedCallback()
+        {
+          @Override
+          public ServerView.CallbackAction serverRemoved(DruidServer server)
+          {
+            if (server.getName().equals("host:8080")) {
+              serverRemovedCalled.countDown();
+              return ServerView.CallbackAction.CONTINUE;
+            } else {
+              throw new RE("Unknown server [%s]", server.getName());
+            }
+          }
+        }
+    );
+
+    httpServerInventoryView.start();
+
+    druidNodeDiscovery.listener.nodesAdded(ImmutableList.of(druidNode));
+
+    initializeCallback1.await();
+    segmentAddLathces.get(segment1.getIdentifier()).await();
+    segmentDropLatches.get(segment1.getIdentifier()).await();
+    segmentAddLathces.get(segment2.getIdentifier()).await();
+    segmentAddLathces.get(segment3.getIdentifier()).await();
+    segmentAddLathces.get(segment4.getIdentifier()).await();
+    segmentDropLatches.get(segment2.getIdentifier()).await();
+
+    DruidServer druidServer = httpServerInventoryView.getInventoryValue("host:8080");
+    Assert.assertEquals(ImmutableMap.of(segment3.getIdentifier(), segment3, segment4.getIdentifier(), segment4),
+                        druidServer.getSegments());
+
+    druidNodeDiscovery.listener.nodesRemoved(ImmutableList.of(druidNode));
+
+    serverRemovedCalled.await();
+    Assert.assertNull(httpServerInventoryView.getInventoryValue("host:8080"));
+
+    EasyMock.verify(druidNodeDiscoveryProvider);
+
+    httpServerInventoryView.stop();
+  }
+
+  private static class TestDruidNodeDiscovery implements DruidNodeDiscovery
+  {
+    Listener listener;
+
+    @Override
+    public Collection<DiscoveryDruidNode> getAllNodes()
+    {
+      throw new UnsupportedOperationException("Not Implemented.");
+    }
+
+    @Override
+    public void registerListener(Listener listener)
+    {
+      listener.nodesAdded(ImmutableList.of());
+      this.listener = listener;
+    }
+  }
+
+  private static class TestHttpClient implements HttpClient
+  {
+    BlockingQueue<ListenableFuture> results;
+    int requestNum = 0;
+
+    TestHttpClient(List<ListenableFuture> resultsList)
+    {
+      results = new LinkedBlockingQueue<>();
+      results.addAll(resultsList);
+    }
+
+    @Override
+    public <Intermediate, Final> ListenableFuture<Final> go(
+        Request request, HttpResponseHandler<Intermediate, Final> httpResponseHandler
+    )
+    {
+      throw new UnsupportedOperationException("Not Implemented.");
+    }
+
+    @Override
+    public <Intermediate, Final> ListenableFuture<Final> go(
+        Request request, HttpResponseHandler<Intermediate, Final> httpResponseHandler, Duration duration
+    )
+    {
+      if (requestNum++ == 0) {
+        //fail first request immediately
+        throw new RuntimeException("simulating couldn't send request to server for some reason.");
+      }
+
+      if (requestNum++ == 1) {
```

Comment:
This should be just requestNum == 1.

Reply:
no, incrementing is necessary so that subsequent requests succeed or else all requests will just keep on failing.

here is what is being simulated...

1st request is failed as if the server rejected even connection setup
2nd reques is as if server responded with error response
subsequent requests succeed with appropriate test expectations.

Reply:
I meant, requestNum should be increased once whenever go() is called. Since it is already incremented above, it looks not to be needed here. You can see that this if clause is never executed.

Reply:
ah, wait, i see ... first check would have incremented it already. you are right.

Reply:
fixed